### DISCUSSION TITLE
Using SynthMorph "shapes" on non-MRI data and variable resolutions

### DISCUSSION THREAD

**Discussion by caacree** [voxelmorph/voxelmorph#490] [Q&A]
Hi, 
I'm looking to use the SynthMorph shapes model to do image registration for multiple images taken at an EM / intracellular scale. I'm new to using DL models, so forgive me if these are basic questions. 

Following the tutorial notebook, I can use the shapes model with the pretrained weights to unwarp my images, but only when scaled down to 256x256 resolution. When I try and use the original resolutions (or (2048, 2048), etc), I get the error `ConcatOp : Dimension 2 in both shapes must be equal: shape[0] = [1,256,32,32] vs. shape[1] = [1,256,256,256]`. Do I need to train a new model from scratch and have the input images all be my desired size instead of 256x256? Or is there some way to use different resolutions than the training data?

Also, we're scanning small sections of tissue samples, which don't have defined boundaries like MRI data and don't always fully overlap. I see that in the Voxelmorph tutorial there is blank padding added to the edges for MRI data, but how would you suggest modifying training data to allow for the moving and fixed images to not be fully overlapping? I could randomly shift or crop the generated warped and/or fixed training images, but wanted your thoughts on if that is a valid strategy. 

Thanks!

**Comment by adalca** [comment]
I don't believe you need to retrain, but It's possible you need to tweak how the model is loaded. @mu40 will know more, so tagging him here. 

Can you post some example images by any chance?

**Comment by mu40** [comment]
While you should not have to retrain this model to use a different image shape, it is hard to predict what went wrong without seeing your code. If the image shape is `(2048, 2048)`, for example, the shape of both model inputs should be `(1, 2048, 2048, 1)`. That is, the model expects a leading batch and a trailing feature dimension. In contrast, we only specify the spatial dimensions `in_shape = (2048, 2048)` when setting up the model in the demo. That being said, you will get best results when the test distribution of image characteristics is included in the training distribution (spatial scales, shapes, etc.).

**Comment by caacree** [comment]
Thanks for the responses! I've attached some example images here and the notebook I'm using; only the last cell contains my code, which lists shape of both inputs as `(1, 2048, 2048, 1)`. However, the code only runs if the shapes are `(1, 256, 256, 1)` (output below). 

<img width="1017" alt="image" src="https://user-images.githubusercontent.com/4749248/201948630-a0892cb4-e94a-418b-b7a0-2c9693f4c480.png">


My goal is to map EM (fixed) and MS (moving) datasets which have been warped, as seen in the examples. You can see that the edges don't perfectly align, the EM data is actually manually cropped in these to roughly contain the MS data. Any tips on how best to tweak training would be appreciated! As mentioned, I'm thinking of shifting / cropping either the fixed or moving images in the label_maps. 

[Colab Notebook](https://colab.research.google.com/drive/1UExeuHyai9DOnRMst0-EY7PeJZBZbwlU?usp=sharing)
[Example Images](https://drive.google.com/drive/folders/1hKaSkGy7IH43qZAJQO45ufxstyUn0DAR?usp=sharing)

**Comment by mu40** [comment]
Thanks for sharing your code and images. The model weights are independent of the input image size. However, you need to specify the image shape you want when constructing the model, by passing `inshape=(2048, 2048)` to `vxm.networks.VxmDense` (cell 6, after which you will of course no longer be able to fit the generator in cell 7).
